src/NuGet.Core/NuGet.PackageManagement/NuGetProject.cs

```diff
@@ -74,13 +78,36 @@ public virtual Task PostProcessAsync(INuGetProjectContext nuGetProjectContext, C
             return Task.FromResult(0);
         }
 
+        public virtual void EnsureMetadataInitialized()
+        {
+            // do nothing by default.
+        }
+
+        private void EnsureInitialized()
+        {
+            lock (_lckObj)
+            {
+                if (!_initialized)
```

Comment:
As mentioned on another comment, more robust way to avoid executing it twice for multi threaded.